cmd/quadlet/main.go

```diff
@@ -0,0 +1,246 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/quadlet/pkg/quadlet"
+	"github.com/containers/quadlet/pkg/systemdparser"
+)
+
+var verboseFlag bool
+var versionFlag bool
+var isUser bool
+
+var noKmsg bool = false
+var kmsgFile *os.File = nil
+
+func logToKmsg(s string) bool {
+	if noKmsg {
+		return false
+	}
+
+	if kmsgFile == nil {
+		f, err := os.OpenFile("/dev/kmsg", os.O_WRONLY, 0644)
+		if err != nil {
+			noKmsg = true
+			return false
+		}
+		kmsgFile = f
+	}
+
+	if _, err := kmsgFile.Write([]byte(s)); err != nil {
+		kmsgFile.Close()
+		kmsgFile = nil
+		return false
+	}
+
+	return true
+}
+
+func Log(format string, a ...any) {
+	s := fmt.Sprintf(format, a...)
+	line := fmt.Sprintf("quadlet-generator[%d]: %s", os.Getpid(), s)
+
+	if !logToKmsg(line) {
+		// If we can't log, print to stderr
+		fmt.Fprintf(os.Stderr, "%s\n", line)
+		os.Stderr.Sync()
+	}
+}
+
+var debugEnabled bool = false
+
+func enableDebug() {
+	debugEnabled = true
+}
+
+func Debug(format string, a ...any) {
+	if debugEnabled {
+		Log(format, a...)
+	}
+}
+
+func getUnitDirs(user bool) []string {
+
+	unitDirsEnv := os.Getenv("QUADLET_UNIT_DIRS")
+	if len(unitDirsEnv) > 0 {
+		return strings.Split(unitDirsEnv, ":")
+	}
+
+	dirs := make([]string, 0)
+	if user {
+		if configDir, err := os.UserConfigDir(); err == nil {
+			dirs = append(dirs, path.Join(configDir, "containers/systemd"))
+		}
+	} else {
+		dirs = append(dirs, quadlet.UnitDirAdmin)
+		dirs = append(dirs, quadlet.UnitDirDistro)
+	}
+	return dirs
+}
+
+func loadUnitsFromDir(sourcePath string, units map[string]*systemdparser.UnitFile) {
+	files, err := os.ReadDir(sourcePath)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			Log("Can't read \"%s\": %s", sourcePath, err)
+		}
+		return
+	}
+
+	for _, file := range files {
+		name := file.Name()
+		if units[name] == nil &&
+			(strings.HasSuffix(name, ".container") ||
+				strings.HasSuffix(name, ".volume")) {
+			path := path.Join(sourcePath, name)
+
+			Debug("Loading source unit file %s", path)
+
+			if f, err := systemdparser.ParseUnitFile(path); err != nil {
+				Log("Error loading '%s', ignoring: %s", path, err)
+			} else {
+				units[name] = f
+			}
+		}
+	}
+}
+
+func generateServiceFile(outputPath string, service *systemdparser.UnitFile) error {
+	Debug("writing '%s'", service.Path)
+
+	service.PrependComment("",
+		"Automatically generated by quadlet-generator",
+		"")
+
+	f, err := os.Create(service.Path)
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	err = service.Write(f)
+	if err != nil {
+		return err
+	}
+
+	f.Sync()
+
+	return nil
+}
+
+func enableServiceFile(outputPath string, service *systemdparser.UnitFile) {
```

Comment:
Note to self: I do not understand what this function does but did not inspect the callees yet.

Reply:
This implements the [Install] section in the .container unit file. This is a section that is otherwise implemented by the "systemctl enable" call. See #20 for the origin of this.